objective-c/README.md

```diff
@@ -0,0 +1,522 @@
+
+`#pragma mark - [main title]`
+
+`#pragma mark [sub-title]`
+
+## Naming global constants
+
+Just like notification names, enum types... etc. They need to reflect to the container class, so 
+it should use it as a base. This help the user of this class to use code completion without checking the classes header.
+
+Example:
+
+	extern NSString * const MYClassDidSomethingNotification;
+
+	extern const CGFloat MYClassConstantFloatValue;
+
+	typedef NS_ENUM(NSInteger, MYClassSide)
+	{
+	MYClassSideLeft,
+	MYClassSidesRight
+	};
+
+	@interface MyClass : ...
+
+
+## Asserts
+
+Use asserts to keep thigt conditions in DEBUG mode. It will help you to find programming mistakes and/or
+invalid states in runtime.
+
+Examples:
+
+	- (BOOL)checkThis:(NSNumber *)number
+	{
+		NSParameterAssert(number);
+		...
+	}
+
+	- (void)myMethod
+	{
+		const NSUInteger numberOfItems = [self numberOfItems];
+		
+		NSAssert(0 < numberOfItems, @"No item has been found");
+		
+		...
+	}
+
+Handy, isn't it?
+
+> Avoid to use messages in the NSAssert body!
+> The asserts are included when you're in DEBUG mode - so the message won't be sent in RELEASE build 
+> which can change it's behaviour.
+
+> Note: NSParameterAssert became deprecated when the _\_Nonnull_ annotation was announced.
+> _- (void)checkThis:(NSNumber * \_Nonnull)number;_
+> This would help you when compiling the app instead of getting exceptions in runtime, 
+> but it's a real pain in the ass to add this to your existing codebase.
+
+## Initializers
+
+Always check the header for designated initializers! Override, or use it in your implementation.
+
+> If you create a new one and your class does't work normally with the
+> default initializer of it's superclass then mark it with `NS_DESIGNATED_INITIALIZER`
+
+Example:
+
+MYCar.h
+
+	@interface MYCar : MYVehicle
+
+	- (instancetype)initWithColor:(UIColor *)color numberOfWheels:(NSUInteger)numberOfWheels NS_DESIGNATED_INITIALIZER;
+	...
+
+	@end
+
+MYCar.m
+
+	@interface MYCar
+
+	@property (nonatomic) NSUInteger numberOfWheels;
+
+	@end
+
+
+	@implementation
+
+	- (instancetype)initWithColor:(UIColor * _Nonnull)color numberOfWheels:(NSUInteger)numberOfWheels
+	{
+		self = [super initWithColor:color];	// designated initializer of MYVehicle class
+		if (self) {
+			self.numberOfWheels = numberOfWheels;
+		}
+
+		return self;
+	}
+
+	...
+
+	@end
+
+## Properties
+
+You need to keep in mind several things when using properties in your class.
+
+### Visibility
+
+If your property is public and read-only, make sure it always has the right value
+(especially in _UIViewController_ subclasses, see below).
+
+### Weak, copy or strong?
+
+In general: 
+
+* Use weak for storing delegates,
+
+* Copy for classes which have mutable versions (NSString, NSArray, NSDictionary...) and blocks,
+
+* Strong for everything else.
+
+### Local variable vs property
+
+Local variables are slowly being dropped from the framework - I recommend to declare properties all the time
```

Comment:
Just to be clear, local variables are being declared by the compiler behind the scenes when you declare an `@property` in most cases.  It isn't if you define an `@property` but override the getter and setter.